CorsixTH/Lua/dialogs/edit_room.lua

```diff
@@ -1086,71 +1114,90 @@ local function checkDoorWalls(x, y, wall, has_swingdoor)
-  if not (tile_flags.buildable or tile_flags.passable or tile_flags.owner == player_id) then return false end
-  if not flag_names then return true end
-  return tile_flags[flag_names[1]] and tile_flags[flag_names[2]]
+  -- check builable and own it
+  if not (tile_flags.buildable or tile_flags.owner == player_id) then return false end
+  -- any object will cause it to be blocked (ignore litter)
+  if not(tile_flags.thob == 0 or tile_flags.thob == 62) then return false end
+  -- check if its passable that no object footprint blocks it
+  if tile_flags.passable then return world:isTileExclusivelyPassable(xpos, ypos, 1) end
+  return true
 end
 
 function UIEditRoom:setDoorBlueprint(orig_x, orig_y, orig_wall)
-  local x = orig_x
-  local y = orig_y
-  local wall = orig_wall
-
-  -- Used to get the adjacent tiles when placing swing doors.
-  local x_mod
-  local y_mod
-  if wall == "south" then
-    y = y + 1
-    wall = "north"
-    x_mod = 2
-  elseif wall == "east" then
-    x = x + 1
-    wall = "west"
-    y_mod = 2
-  elseif wall == "north" then
-    x_mod = 2
-  else
-    y_mod = 2
+  --! Calculate position offsets and door blueprint wall values
+  --! param x - doors blueprint x value
+  --! param y - doors blueprint y value
+  --! param wall - original wall orientation
+  --! return x - updated x value
+  --! return y - updated y value
+  --! return x_mod - offest value to apply to tile count to determine relative position
+  --! return y_mod - offset value to apply to tile count to determine relatitve position
+  --! return wall - wall orientation style (only 2 styles)
+  local function doorWallOffsetCalculations(x, y, wall)
+    local x_mod
+    local y_mod
+    if wall == "south" then
+      y = y + 1
+      wall = "north"
+      x_mod = 2
+    elseif wall == "east" then
+      x = x + 1
+      wall = "west"
+      y_mod = 2
+    elseif wall == "north" then
+      x_mod = 2
+    else
+      y_mod = 2
+    end
+    return x, y, x_mod, y_mod, wall
   end
+
+  local x, y, x_mod, y_mod, wall = doorWallOffsetCalculations(orig_x, orig_y, orig_wall)
   local map = TheApp.map.th
 
   if self.blueprint_door.anim then
     if self.room_type.swing_doors then
       if self.blueprint_door.anim[1] then
+        -- retrieve the old door position details to reset the blue print
+        local oldx, oldy, oldx_mod, oldy_mod, old_wall = doorWallOffsetCalculations(self.blueprint_door.floor_x,
+          self.blueprint_door.floor_y, self.blueprint_door.wall)
         -- If we're dealing with swing doors the anim variable is actually a table with three
         -- identical "doors".
         for i, anim in ipairs(self.blueprint_door.anim) do
           anim:setAnimation(self.anims, self.blueprint_door.old_anim[i],
             self.blueprint_door.old_flags[i])
           anim:setTag(nil)
           self.blueprint_door.anim[i] = nil
+          oldx = oldx_mod and self.blueprint_door.floor_x + (i - oldx_mod) or self.blueprint_door.floor_x
```

Comment:
Why did you put this inside the anim loop? It looks like it would do the same thing every loop iteration unless I'm missing something.

Reply:
Just to cover this off - I've updated the code to support the 3 tile blueprint for swing doors.

The doorWallOffsetCalculations function returns nil for oldx_mod/oldy_mod depending on orientation and the (i -oldx(y)_mod) calculation performed within.

The old code would end that loop and then call
map:setCell(self.blueprint_door.floor_x, self.blueprint_door.floor_y, 4, 24) which would be the centre tile of the door.

But that is just one blueprint tile being reset. So upon adding the support for the blueprint for the entire 3 tiles of the swing door, all 3 tiles need to be cleaned up, and the supporting calculation is done on the orientation.